### ISSUE TITLE
Program gets killed when running Ginex

### ISSUE THREAD

**Issue by Liu-rj** [SNU-ARC/Ginex#3]
Hi,
Following your commands in README, when I ran
```
sudo PYTHONPATH=/opt/conda/envs/npc/lib/python3.9/site-packages cgexec -g memory:8gb /opt/conda/envs/npc/bin/python -W ignore run_ginex.py --neigh-cache-size 6000000000 --feature-cache-size 6000000000 --sb-size 1500
```
The program prints the following:
```
Preparing dataset...
Done!

==============================
Running Epoch 0...
Running 0th superbatch of total 1 superbatches                                                                                            
Step 1: Superbatch Sample                                                                                                                 
Epoch 00:   0%|                                                                                               | 0/1207179 [00:00<?, ?it/s]
[1]    55263 killed     sudo PYTHONPATH=/opt/conda/envs/npc/lib/python3.9/site-packages cgexec -g   -
```
As is printed, the program gets killed. And I notice that there is already an issue reporting this  #1, but I don't find a solution in that discussion.

My build environment (AWS EC2 g4dn.metal instance):
```
cpu_info: Intel(R) Xeon(R) Platinum 8259CL CPU @ 2.50GHz
gpu_info: NVIDIA Tesla T4 16GB PCIe
memory: 8g cgroup limit used
swapfile:4g used
cuda: 12.0
python: 3.9.16
pytorch: 2.0.0
nvidia-driver: 525.85.12
Ubuntu: 20.04.1
```

Any help or guidance is appreciated, thanks

**Comment by Liu-rj** [comment]
Also, I can successfully run the baseline following:
```
sudo PYTHONPATH=/opt/conda/envs/npc/lib/python3.9/site-packages cgexec -g memory:8gb /opt/conda/envs/npc/bin/python -W ignore run_baseline.py
```
The output is like:
```
Epoch 00: 100%|██████████████████████████████████████████████████████████████████████████████▉| 1206179/1207179 [08:34<00:00, 2342.82it/s]
Epoch 00, Loss: 1.8444, Approx. Train: 0.4852
Epoch time: 514847.6493 ms
```

**Comment by ilil96** [comment]
Can you check why the program gets killed by looking at the dmesg output?

**Comment by Liu-rj** [comment]
I change the `num_workers` and `prefetch_factor` of `GinexNeighborSampler` to 32 and 2, respectively. And now I can run the first epoch successfully but still get killed at the 2nd epoch. I also tried even less workers, for example 4, and this time the program gets killed at 3rd epoch. I believe there is memory leak  somewhere in the code.

The dmesg gives
```
[101093.587436] oom-kill:constraint=CONSTRAINT_MEMCG,nodemask=(null),cpuset=/,mems_allowed=0-1,oom_memcg=/8gb,task_memcg=/8gb,task=python,pid=459713,uid=0
[101093.592404] Memory cgroup out of memory: Killed process 459713 (python) total-vm:182849844kB, anon-rss:5955836kB, file-rss:383304kB, shmem-rss:20kB, UID:0 pgtables:145560kB oom_score_adj:0
```

**Comment by ilil96** [comment]
That's strange. I will try to reproduce the bug by myself and get back to you.

**Comment by WWWzq-01** [comment]
Hello, I also encountered a similar issue. Have you resolved it now? @Liu-rj

**Comment by Liu-rj** [comment]
I chose to not use cgroup for the memory limit, so this problem just gone... But there indeed is a bug in the sampling code that will cause core dump. I don't know if it's also the cause of this problem. Anyway I'll make a PR for it and you can see if it works to you. @WWWzq-01

**Comment by WWWzq-01** [comment]
> I chose to not use cgroup for the memory limit, so this problem just gone... But there indeed is a bug in the sampling code that will cause core dump. I don't know if it's also the cause of this problem. Anyway I'll make a PR for it and you can see if it works to you. @WWWzq-01

Thank u.I chose to not use cgroup for the memory limit too.Then the problem was solved.

**Comment by Liu-rj** [comment]
FYI. The PR is here #10.

**Comment by Liu-rj** [comment]
So I'll just close this issue.

**Comment by WWWzq-01** [comment]
Hello,I have merged PR #10, but the program is still getting killed.
the `dmesg` gives:
```
[2423580.910294] oom-kill:constraint=CONSTRAINT_MEMCG,nodemask=(null),cpuset=/,mems_allowed=0-1,oom_memcg=/8gb,task_memcg=/8gb,task=python3.6,pid=1627997,uid=0
[2423580.912078] Memory cgroup out of memory: Killed process 1627997 (python3.6) total-vm:107600208kB, anon-rss:3980652kB, file-rss:1256152kB, shmem-rss:12296kB, UID:0 pgtables:32664kB oom_score_adj:0
[2423581.092390] oom_reaper: reaped process 1627997 (python3.6), now anon-rss:0kB, file-rss:108788kB, shmem-rss:12296kB
```
@Liu-rj
